pandas/core/case_when.py

```diff
@@ -0,0 +1,217 @@
+
+
+def case_when(
+    *args: tuple[tuple],
+    default=lib.no_default,
+    level: int | None = None,
+) -> Series:
+    """
+    Replace values where the conditions are True.
+    It is a convenience wrapper around `pd.Series.mask`,
+    and is handy for multiple conditions and replacements.
+
+    Parameters
+    ----------
+    *args : array-like, scalar
+        Variable argument of tuples of conditions and expected replacements.
+        Takes the form:  `(condition0, replacement0)`,
+            `(condition1, replacement1)`, ... .
+        `condition` should be a 1-D boolean array.
+        When multiple boolean conditions are satisfied,
+        the first replacement is used.
+
+    default : scalar, array-like, default None
+        If provided, it is the replacement value to use
+        if all conditions evaluate to False.
+        If not specified, entries will be filled with the
+        corresponding NULL value (``np.nan`` for numpy dtypes,
+        ``pd.NA`` for extension dtypes).
+
+    level : int, default None
+        Alignment level if needed.
+
+        .. versionadded:: 2.2.0
+
+    Returns
+    -------
+    Series
+
+    See Also
+    --------
+    Series.mask : Replace values where the condition is True.
+
+    Examples
+    --------
+    >>> df = pd.DataFrame({
+    ...     "a": [0,0,1,2],
+    ...     "b": [0,3,4,5],
+    ...     "c": [6,7,8,9]
+    ... })
+    >>> df
+       a  b  c
+    0  0  0  6
+    1  0  3  7
+    2  1  4  8
+    3  2  5  9
+
+    >>> pd.case_when((df.a.gt(0), df.a),   # condition, replacement
+    ...              (df.b.gt(0), df.b),   # condition, replacement
+    ...              default=df.c)        # optional
+    0    6
+    1    3
+    2    1
+    3    2
+    Name: c, dtype: int64
+    """
+    from pandas import Series
+    from pandas._testing.asserters import assert_index_equal
+
+    # summary of logic -
+    # get booleans and replacements
+    # and iteratively get the final output via Series.mask
+    # iteration starts from the end,
+    # ensuring that if there are any booleans with the same replacement,
+    # the first one is taken
+    # as much as possible the computation is pushed to Series.mask
+    if not len(args):
+        raise ValueError(
+            "Kindly provide at least one boolean condition, "
+            "with a corresponding replacement."
+        )
+
+    for num, entry in enumerate(args):
+        if not isinstance(entry, tuple):
+            raise TypeError(
+                f"Argument{num} should be a tuple, instead, got {type(entry)}."
+            )
+        if len(entry) != 2:
+            raise ValueError(
+                f"Argument{num} should have a condition "
+                "and a replacement - it should have a length 2; "
+                f"however, it has a length {len(entry)}."
+            )
+    conditions, replacements = zip(*args)
+    common_dtype = [infer_dtype_from(replacement)[0] for replacement in replacements]
```

Comment:
nit: can you call this `common_dtypes`. This way we don't have a variable that is at one point a list becomes a dtype (which can be confusing).